Fix part of #4030 : Implementing dark mode to various screens.

ayush0402 opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fix part of #4030 
Link to the original PR having discussions over the decided approach : [link](https://github.com/oppia/oppia-android/pull/3951)

Summary of the approach followed in this PR : 
- Introduced new colors in `color_defs.xml` following decided snake_case naming convention. `colors_def.xml` should strictly contain actual color name irrespective of their intended use with their hex color code declaration.
```xml
<color name="lime_green">#90EE90</color>
```
- Then `color_palette.xml` (day/night) splits the colors (to be used by `component_colors.xml`) according to chosen theme. `color_palette.xml` should not contain any name tied to the intended component to be used on, rather should have generic naming of the colors.
> Don't :
```xml
 <color name="add_profile_background_color">@color/oppia_background_black</color>
 <color name="text_input_layout_error_color">@color/error_red</color>
```
> Do : 
```xml
 <color name="background_color">@color/oppia_background_black</color>
 <color name="error_color">@color/error_red</color>
```
- `component_colors.xml` then only references `color_palette.xml`. It uses UI component specific names. Component colors should be shared very little outside of their respective views/fragments/activities.
```xml
  <color name="shared_text_input_edit_text_cursor_color">@color/primary_text_color</color>
  <color name="shared_activity_toolbar_color">@color/toolbar_color</color>
  <!--Styles.xml-->
  <color name="shared_text_input_layout_text_color">@color/primary_text_color</color>
  <color name="shared_input_interaction_edit_text_text_color">@color/primary_text_color</color>
  <color name="shared_text_input_layout_background_color">@color/text_input_background_color</color>
  <!--Admin Auth Activity-->
  <color name="admin_auth_secondary_text_color">@color/description_text_color</color>
  <color name="admin_auth_layout_background_color">@color/background_color</color>
  <!--Add Profile Activity-->
  <color name="add_profile_activity_label_text_color">@color/primary_text_color</color>
  <color name="add_profile_activity_switch_text_color">@color/dark_text_color</color>
```
- `colors_migrating.xml` contains color declarations which are supposed to be in `color_defs.xml` but has not been renamed yet to have actual color name instead of names linked to their use and components. This is a temporary measure to make sure other 4 color files follows the convention decided for them.

Note: colors declared in `component_colors.xml` and `color_palette.xml` should not use "oppia_" prefix to avoid redundancy and must have `_color` suffix. And just opposite applies to `color_defs.xml`


Link to [comment](https://github.com/oppia/oppia-android/pull/4032#discussion_r770286256) by Ben on this PR summarising the approach.

#### Screenshots : 

| Light | Dark |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/76056229/147963117-1edcf8c3-56bf-4ea8-be5d-76bfece4faa8.png" alt="light" width="250"> | <img src="https://user-images.githubusercontent.com/76056229/148102291-19cace40-dc94-4c1b-9807-05d9e32d997a.png" alt="dark" width="250">|
| <img src="https://user-images.githubusercontent.com/76056229/147963324-b64d3fa0-1818-474f-91a0-b92812c94c84.png" alt="light" width="250"> | <img src="https://user-images.githubusercontent.com/76056229/148036805-1d15f279-8313-4fa2-ac34-6c93c2192fc6.png" alt="dark" width="250">|
| <img src="https://user-images.githubusercontent.com/76056229/148102457-af390ad3-7711-4aa3-b5fa-2224c4fbb339.png" alt="light" width="250"> | <img src="https://user-images.githubusercontent.com/76056229/148102508-b043f0c0-50ec-4963-8c3e-cb1ec07e9860.png" alt="dark" width="250">|
| <img src="https://user-images.githubusercontent.com/76056229/148105548-e090c778-bb6e-43d5-9a8c-4ffa6f5cc7eb.png" alt="light" width="250"> | <img src="https://user-images.githubusercontent.com/76056229/148105590-59a3c328-b3df-48a3-9bb7-60a980205318.png" alt="dark" width="250">|

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

ayush0402 committed:
updated till line 10

ayush0402 committed:
line 16

ayush0402 committed:
updated till line 42

ayush0402 committed:
...

ayush0402 committed:
done

ayush0402 committed:
fixed unresolved reference for highlighterDeveloperOptionsNavMenuItem

ayush0402 committed:
highlightedNavMenuItem

ayush0402 committed:
renamed colors.xml to color_defs.xml

ayush0402 committed:
added basic colors

ayush0402 committed:
resolving merge conflict

ayush0402 committed:
Merge branch 'develop' of github.com:oppia/oppia-android into dark_mode_files

ayush0402 committed:
resolved merge conflict

ayush0402 committed:
working on add profile activity

ayush0402 committed:
using style/TextInputLayout to change boxBackground

ayush0402 commented:
@BenHenning I will be implementing dark mode to some layouts in this PR so I just wanted to confirm the approach by which I did it for one layout (Add profile activity) if this is good then I will be pushing more layouts similarly.
| Light | Dark |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/76056229/145724285-ac854a1d-6e52-45a0-b706-f5e9bcdf92c9.png" alt="light" width="250"> | <img src="https://user-images.githubusercontent.com/76056229/145724289-90c5703d-77f4-417f-8fdd-cc3b971c98ab.png" alt="dark" width="250">|

Also I had two problems with this...
- I am using a different primary color as per the provided mocks but by default the Action bar is set to primary color only. I tried using [this](https://stackoverflow.com/a/9249702/14822653) for changing the color of Action bar like...
(Color of action bar is same in both light and dark themes).
```xml
  <!--ACTION BAR STYLE-->
  <style name="OppiaActionBar" parent="Widget.AppCompat.Light.ActionBar">
    <item name="android:background">@color/oppia_action_bar_color</item>
  </style>
```
```xml
  <style name="OppiaTheme" parent="Theme.MaterialComponents.Light.DarkActionBar.Bridge">
    <item name="actionBarStyle">@style/OppiaActionBar</item>
  </style>
```
but this is not working.

- Another issue was about TextInputLayout/TextInputEditText, I am unable to find which attribute for the component is responsible for color of the field name when it is not selected as shown in this video.

https://user-images.githubusercontent.com/76056229/145724990-b47baf33-f792-4465-8455-1b87647f1bdd.mp4

Please let me know if you have any suggestions for these.

ayush0402 committed:
updated color for edit text cursor

BenHenning commented:
Apologies @ayush0402, this will probably take me a bit to review so I won't be able to look at it until tomorrow at the earliest.

BenHenning commented:
@ayush0402 regarding your issues:

1) I know that certain system components (such as the status bar) use different attributes across different SDK versions. You might need to continue searching to figure out which colors correspond to the action bar, and might even need to look at the source code for action bar itself. This isn't always straightforward, and this might actually be one of the trickier cases to solve in the project.

2) Same for text. You should look into the source code for the material component that we're using to see which attributes it uses for color selection for different parts of the layout.

I will be following up with high-level thoughts on the code soon, but I wanted to follow up on these points, first.

BenHenning commented on app/src/main/res/layout/add_profile_activity.xml:
```diff
@@ -28,7 +28,7 @@
         android:layout_width="match_parent"
         android:layout_height="match_parent"
-        android:background="@color/add_profile_background">
+        android:background="@color/oppia_add_profile_background_color">
```
A first thought: we probably don't need to prefix *everything* with 'oppia_'. I can see it being useful for custom colors (i.e. in color_defs), but everywhere else it's redundant.

ayush0402 replied:
Done.

BenHenning commented on app/src/main/res/layout/add_profile_activity.xml:
```diff
@@ -71,7 +71,7 @@
             android:contentDescription="@string/required_fields_content_description"
             android:text="@string/add_profile_required"
-            android:textColor="@color/oppia_primary_text"
+            android:textColor="@color/oppia_primary_text_color"
```
'primary_text_color' seems too generic here and is more like a color_palette color than a color component one. A color component might be more like 'label_text_color'. In many cases, the style might give a clue on how to name the color.

ayush0402 replied:
Done.

BenHenning commented on app/src/main/res/values-night/color_palette.xml:
```diff
@@ -0,0 +1,13 @@
+  <color name="accent_color">@color/oppia_accent_2</color>
+  <color name="primary_text_color">@color/oppia_white_subtext</color>
+  <color name="heading_text_color">@color/white</color>
+  <color name="add_profile_background_color">@color/oppia_background_black</color>
+  <color name="switch_description_color">@color/oppia_white_subtext</color>
+  <color name="edit_text_input_background">@color/oppia_edit_text_background</color>
+  <color name="add_profile_switch_text_color">@color/white</color>
+  <color name="text_input_layout_error_color">@color/error_red</color>
```
These are component colors (they're tied to specific components) and thus shouldn't be defined here.

ayush0402 replied:
Done.

BenHenning commented on app/src/main/res/values/color_defs.xml:
```diff
@@ -146,4 +146,12 @@
   <color name="coming_soon_start">#7659B6</color>
+  <!--DARK MODE-->
+  <color name="oppia_white_subtext">#C4C4C4</color>
+  <color name="oppia_primary_2">#3bd1c4</color>
```
We should list the actual color here, not 'primary' (that's more like color_palette).

ayush0402 replied:
Done.

BenHenning commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <color name="oppia_primary_color">@color/primary_color</color>
```
Some of these just repeat variants in color_palette. That's not the intention of component_colors.

Think of it this way:
- color_defs defines the actual hex colors (only used by color_palette)
- color_palette defines a list of generic colors that are used everywhere through component colors (only used by component_colors)
- component_colors defines more specific colors (either colors only used by a single component, similar to strings & dimensions, or generic colors that don't need to be switched out by night mode)

ayush0402 replied:
Removed from component_colors

BenHenning reviewed (dismissed):
Thanks @ayush0402! And, apologies for the delay. I left some high-level thoughts in comments.

Overall, I think this is on the right approach but there may still be some confusion around the differences between the different color files. I also suspect you might benefit from adding a few more examples (especially layouts since this results in more component color cases which I expect is the more confusing one to figure out). Beyond that, it might be good to wait to finish the PR in its full until Rajat has a chance to review this (though that will mean waiting 1+ weeks for review). It's up to you if you want to make more progress in the meantime, just know that a lot of it might have to be redone if he has more major feedback later.

BenHenning commented:
Hi. As of today, some main reviewers have taken time off for the next few weeks, so it may take a little while before we can look at this PR. We appreciate your patience while some of our team members recharge. We'll be fully returning on 4 January 2021.

ayush0402 commented:
Thanks, I figured out the issue with changing colors for the "ActionBar". Actually we were not using Action Bar at all for that screen, It was using a NoActionBar variant of the theme and it was actually a custom toolbar defined under layouts/toolbar.xml which was being for various layouts. Changing its properties under layouts/toolbar.xml works for changing the color other than colorPrimary now.

ayush0402 committed:
added color under toolbar and removing generic colors from component_color.xml

ayush0402 committed:
fixed namings of new colors in color_defs.xml

ayush0402 committed:
renamed oppia_heading_text_color

ayush0402 committed:
fixing color naming in component_colors.xml and color_palette.xml

ayush0402 committed:
fixing naming in component_colors.xml and color_palette.xml

ayush0402 committed:
fixing naming in component_colors.xml and color_palette.xml

ayush0402 committed:
fixed namings in component_colors.xml and color_palette.xml

ayush0402 committed:
Merge branch 'develop' of github.com:oppia/oppia-android into dark_mode_files

 Conflicts:
	app/src/main/res/layout/add_profile_activity.xml

ayush0402 committed:
resolved merge conflict

ayush0402 committed:
admin_auth_activity

ayush0402 committed:
working on admin_pin_activity

ayush0402 committed:
fixed color for TextInputLayout

ayush0402 committed:
changed administrator_controls

ayush0402 committed:
changed statusBar color

ayush0402 committed:
fixed action bar color in OppiaTheme

ayush0402 committed:
nit changes

ayush0402 commented:
@rt4914 @BenHenning I implemented dark mode for 4 layouts right now. PTAL. And should I add more screen to the current PR or proceed further in another PR? I added screenshots to the PR description [here](https://github.com/oppia/oppia-android/pull/4032#issue-1075945982).

ayush0402 committed:
renamed heading_text_color

rt4914 reviewed (changes requested):
Just reviewed couple of files but for this I will need to go to previous rules that we defined and messages so will take some more time to do full review.

BenHenning commented:
@rt4914 deferring to you for the rest of this review. Please let me know if my thoughts are needed on anything.

ayush0402 committed:
Merge branch 'develop' of github.com:oppia/oppia-android into dark_mode_files

rt4914 commented on app/src/main/res/drawable/color_cursor.xml:
```diff
@@ -1,5 +1,5 @@
 <shape xmlns:android="http://schemas.android.com/apk/res/android">
   <size android:width="0.5dp" />
-  <solid android:color="@color/oppia_primary_text" />
+  <solid android:color="@color/label_text_color" />
```
I think instead of using `label_text_color` define another component color named `cursor_color` and use that instead as it would be add more clear context.

ayush0402 replied:
Yeah, that will be better. Done.

rt4914 commented on app/src/main/res/layout-land/admin_auth_activity.xml:
```diff
@@ -80,7 +80,7 @@
             android:layout_marginTop="12dp"
             android:text="@string/admin_auth_sub"
-            android:textColor="@color/oppia_admin_auth_secondary_text"
+            android:textColor="@color/admin_auth_secondary_text"
```
One major suggestion. I think all those colors which are defined in `component_colors.xml` should have a end suffix of `_color` and everything inside `color_palette` and `color_defs` should not have suffix `_color` this way we can easily add a regex for all three files to clearly follow above rules and also it will set a good coding convention.

rt4914 replied:
@BenHenning Any thoughts on this?

rt4914 replied:
Basically we need a convention to make sure when to use `_color` as suffix and when not too.

BenHenning replied:
I think that makes sense for color_defs and component_colors, but how would it work for color_palette? Could you maybe add some examples to demonstrate the convention?

rt4914 replied:
Actually you are right. @BenHenning 
We can need `_color` as suffix for all values in `color_palette`.
So lets follow this

`color_defs` -> does not contain `_color` as in this file we are mentioning colors like `black`, `white` which by default indicate color.
`color_palette` -> all values contain `_color` as suffix
`color_components` -> all values contain `_color` as suffix


@BenHenning Any thoughts?

ayush0402 replied:
This convention sounds good to me.

BenHenning replied:
That SGTM @rt4914.

rt4914 commented on app/src/main/res/layout-land/administrator_controls_account_actions_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_account_actions_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
This color also does not add context.  Something like `sub_heading_color` might be more clear or `adminsitrator_controls_sub_heading_color`

ayush0402 replied:
Changed to `admin_controls_sub_heading_color`.

rt4914 commented on app/src/main/res/layout-land/administrator_controls_app_information_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_app_information_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout-land/administrator_controls_download_permissions_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_download_permissions_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout-land/administrator_controls_fragment.xml:
```diff
@@ -13,7 +13,7 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent"
-    android:background="@color/mid_grey_30"
+    android:background="@color/admin_controls_background"
```
`admin_controls_background_color`

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout-land/administrator_controls_general_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_general_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout-land/administrator_controls_profile_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_profile_management_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout-sw600dp/administrator_controls_account_actions_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_account_actions_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout-sw600dp/administrator_controls_app_information_view.xml:
```diff
@@ -27,7 +27,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_app_information_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout-sw600dp/administrator_controls_download_permissions_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_download_permissions_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout-sw600dp/administrator_controls_general_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_general_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout/administrator_controls_account_actions_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_account_actions_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout/administrator_controls_app_information_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_app_information_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout/administrator_controls_download_permissions_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_download_permissions_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout/administrator_controls_general_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_general_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout/administrator_controls_profile_view.xml:
```diff
@@ -23,7 +23,7 @@
       android:paddingBottom="20dp"
       android:text="@string/administrator_controls_profile_management_label"
-      android:textColor="@color/oppia_primary"
+      android:textColor="@color/primary_color"
```
Ditto

ayush0402 replied:
Done.

rt4914 reviewed (changes requested):
@ayush0402 Really nice work.

rt4914 commented:
@BenHenning Can you have a look at this comment and leave your thoughts.
https://github.com/oppia/oppia-android/pull/4032#discussion_r786255540

ayush0402 committed:
Merge branch 'develop' of github.com:oppia/oppia-android into dark_mode_files

ayush0402 committed:
made requested changes

BenHenning commented on app/src/main/res/values/color_palette.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <color name="primary_color">@color/oppia_primary</color>
```
Shouldn't this point to a specific color rather than oppia_primary? Ditto for the others.

ayush0402 replied:
Yes, but the reason I didn't change it yet is because most activities(other than the ones I changed in this PR) are still referencing to older pre-defined colors in `colors_def.xml` like (oppia_primary in color_defs.xml). I thought it will be easier to track the changes if we change those default values after we have implemented dark mode through `component_colors.xml` to most of the layouts. As then we will only have to change the values in `color_palette.xml` rather than many files.
@BenHenning What are your views on this? If you suggest otherwise I will rename existing colors in `color_defs.xml` in this PR only.

BenHenning replied:
Might it be possible to introduce a temporary fourth file, maybe colors_migrating.xml that have *all* colors that need to be categorized and/or renamed? That would allow you to ensure each of the other color files is setup exactly the way we want it to be long-term without having to make any concessions.

ayush0402 replied:
Sounds good. Added `colors_migrating.xml`.

BenHenning replied:
Thanks. Can this file at least be updated to point to the correct color? It doesn't really make sense for the new files to reference colors_migrating.xml since only migrated colors should go in these files.

BenHenning commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <color name="label_text_color">@color/primary_text_color</color>
```
This is rather generic and feels like something that would go in color_palette, not component_colors. Ditto for the next few.

I'm actually thinking that component colors should be shared very little outside of their respective views/fragments/activities.

ayush0402 replied:
I tried restructuring `component_colors` into something more restricted into the scope of their activities/views only. PTAL now. If something is still off can you please provide examples for how it should be.

BenHenning reviewed (commented):
Left a few follow-up thoughts.

rt4914 commented:
@BenHenning PTAL at this https://github.com/oppia/oppia-android/pull/4032#discussion_r788049699

ayush0402 committed:
replaced label_text_color with more specific component colors

ayush0402 committed:
restructuring component_colors restricting colors to their activities/views

ayush0402 committed:
renamed colors in component_colors.xml

ayush0402 commented:
@BenHenning PTAL now.

BenHenning commented on app/src/main/res/values/color_defs.xml:
```diff
@@ -146,4 +145,13 @@
   <color name="coming_soon_end">#674172</color>
   <color name="coming_soon_start">#7659B6</color>
+  <!--DARK MODE-->
+  <color name="oppia_subtext_white">#C4C4C4</color>
```
'subtext' feels like a palette detail, ditto for background & container. Can we describe the color directly in its name, instead?

ayush0402 replied:
Done.

BenHenning commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <color name="text_input_edit_text_cursor_color">@color/primary_text_color</color>
+  <color name="toolbar_color">@color/oppia_primary</color>
+  <color name="status_bar_color">@color/oppia_primary_dark</color>
```
These still seem quite generic. I'd expect them to be part of the color palette, and then have per-activity versions like home_activity_toolbar_color.

ayush0402 replied:
Fixed it.

BenHenning reviewed (commented):
Left more follow-up thoughts. PTAL @ayush0402.

ayush0402 committed:
fixed toolbar and statusbar colors in component_colors.xml

ayush0402 committed:
renamed newly added colors in color_defs

ayush0402 committed:
introduced colors_migrating.xml

ayush0402 committed:
shifted accessible grey to colors_migrating.xml

ayush0402 committed:
reverted last commit

ayush0402 committed:
Merge branch 'develop' of github.com:oppia/oppia-android into dark_mode_files

ayush0402 commented:
@BenHenning PTAL.

BenHenning commented on app/src/main/res/values/colors_migrating.xml:
```diff
@@ -0,0 +1,113 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
```
I suggest adding a top-level comment explaining what colors should go in this file. We also ought to add a CODEOWNERS rule and make you & Rajat the codeowners specifically for this file so that people don't add to it (otherwise it'll make the migration harder).

ayush0402 replied:
PTAL.

BenHenning commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <color name="text_input_edit_text_cursor_color">@color/primary_text_color</color>
+  <color name="toolbar_layout_background_color">@color/oppia_primary</color>
+  <!--Themes.xml-->
+  <color name="oppia_themes_status_bar_color">@color/oppia_primary_dark</color>
+  <color name="oppia_action_bar_theme_background_color">@color/oppia_primary</color>
+  <!--Styles.xml-->
+  <color name="text_view_text_color">@color/primary_text_color</color>
+  <color name="text_view_heading_text_color">@color/highlighted_text_color</color>
+  <color name="text_input_layout_text_color">@color/primary_text_color</color>
+  <color name="text_input_layout_stroke_color">@color/primary_text_color</color>
+  <color name="text_input_edit_text_text_color">@color/primary_text_color</color>
+  <color name="input_interaction_edit_text_text_color">@color/primary_text_color</color>
+  <color name="text_input_layout_background_color">@color/text_input_background</color>
+  <color name="text_input_layout_error_color">@color/error_color</color>
```
I think I had a comment for these at one point, but I can't find it. These are too generic for component colors and feel closer to color_palette where this file should have more specific ones (like home_activity_toolbar_color or answer_box_error_color).

ayush0402 replied:
These colors are not used by any components or any layout file directly. Like `toolbar_layout_background_color` is only used in `toolbar.xml` which is in turn included in other layout files. Similar is with other colors as they are only used once in some styles defined in `styles.xml` which is then in turn used for other layouts. So I don't think these can be named any more specific than the styles they are used for. Other option can be to move these to `color_palette.xml` as it is if that makes more sense than keeping them in `component_colors.xml`. So, should I move them to color_palette instead?

BenHenning replied:
Ah, that's interesting. Hmm, I think this needs to stay as a component color but perhaps we can make it clearer that it's not specific to any particular screen, e.g.: ``shared_activity_toolbar_color``. That color could still point to a color_palette ``toolbar_color`` for theming. Any thoughts @rt4914?

My main thinking is that it's probably more important to ensure that only component colors are used in layouts, and are as specific as possible, and that they tie back to color_palette so that day/night mode works easily.

rt4914 replied:
I thinking prefixing `shared_` to all such cases is a nice solution.

ayush0402 replied:
Done.

BenHenning reviewed (commented):
Thanks @ayush0402. The new migration file looks good. Added a couple of thoughts.

ayush0402 committed:
Added comment for colors_migrating.xml

ayush0402 committed:
added myself as CODEOWNER of colors_migrating.xml

rt4914 commented on .github/CODEOWNERS:
```diff
@@ -56,6 +56,7 @@ gradlew.bat @BenHenning
 # All resource files.
 /app/src/main/res/**/*.xml @rt4914
 /utility/src/main/res/**/*.xml @rt4914
+/app/src/main/res/values/colors_migrating.xml @ayush0402
```
@ayush0402 Please also add me to this file.

ayush0402 replied:
Is this correct now?

rt4914 commented on app/src/main/res/layout-sw600dp/administrator_controls_profile_view.xml:
```diff
@@ -37,14 +37,14 @@
       style="@style/Body"
       android:layout_width="match_parent"
-      android:background="@{commonViewModel.selectedFragmentIndex == viewModel.itemIndex? @color/admin_highlighted : @color/white}"
+      android:background="@{commonViewModel.selectedFragmentIndex == viewModel.itemIndex? @color/admin_highlighted : @color/admin_controls_options_background_color}"
```
Can we change this color too: `admin_highlighted`? name it more correctly.

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/layout/add_profile_activity.xml:
```diff
@@ -105,6 +105,7 @@
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
             android:layout_marginTop="@dimen/add_profile_pin_checkbox_margin_top"
+            android:textColor="@color/add_profile_activity_label_text_color"
```
Why was this attribute added, as earlier it was not available?

ayush0402 replied:
That is needed because otherwise the text along with the check-box stays black even in dark-mode.

rt4914 replied:
Ohh okay got it.

rt4914 commented on app/src/main/res/values/color_defs.xml:
```diff
@@ -51,34 +22,9 @@
-  <color name="highlighted_nav_menu_item">#BE563C</color>
-  <color name="highlighted_developer_options_nav_menu_item">#00609C</color>
-  <color name="navbar_header_background">#264653</color>
   <!-- SUBMIT BUTTON INACTIVE -->
```
I think we can remove these comments.

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -0,0 +1,36 @@
+<resources>
+  <color name="text_input_edit_text_cursor_color">@color/primary_text_color</color>
+  <color name="toolbar_layout_background_color">@color/oppia_primary</color>
+  <!--Themes.xml-->
```
Nit spacing before and after text
`<!-- Themes.xml -->`

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -0,0 +1,36 @@
+  <!--Themes.xml-->
+  <color name="oppia_themes_status_bar_color">@color/oppia_primary_dark</color>
+  <color name="oppia_action_bar_theme_background_color">@color/oppia_primary</color>
+  <!--Styles.xml-->
```
Ditto

ayush0402 replied:
Done.

rt4914 commented on app/src/main/res/values/component_colors.xml:
```diff
@@ -0,0 +1,36 @@
+  <color name="input_interaction_edit_text_text_color">@color/primary_text_color</color>
+  <color name="text_input_layout_background_color">@color/text_input_background</color>
+  <color name="text_input_layout_error_color">@color/error_color</color>
+  <!--Admin Auth Activity-->
```
Ditto spacing

ayush0402 replied:
Done.

rt4914 reviewed (changes requested):
@ayush0402 Suggested nit changes.

ayush0402 committed:
requested nit changes

ayush0402 committed:
renamed admin_highlighted in admin_controls

BenHenning commented:
Codeowners LGTM. Please assign back once this PR is considered final from the colors perspective, and I'll take another pass.

rt4914 reviewed (approved):
LGTM, thanks.

ayush0402 committed:
added shared_ prefix to name of shared component_colors

ayush0402 committed:
renamed component_colors

ayush0402 committed:
nit changes

ayush0402 committed:
Merge branch 'develop' of github.com:oppia/oppia-android into dark_mode_files

ayush0402 committed:
nit changes

ayush0402 commented:
@BenHenning PTAL.

BenHenning commented on app/src/main/res/values/colors_migrating.xml:
```diff
@@ -0,0 +1,117 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+  <!-- contains color declarations which are supposed to be in [color_defs.xml] but has not been
+   renamed yet to have actual color name instead of names linked to their use and components.
+   This is a temporary measure to make sure other 4 color files follows the convention decided
+   for them. -->
```
```suggestion
  <!-- Color declarations which are supposed to be in color_defs.xml but have not yet been
   renamed to have actual color name instead of names linked to their use and components.
   This is a temporary measure to make sure other 4 color files follows the convention decided
   for them. -->
```

Some fixes in the beginning, but content wise: maybe it would be clearer just to say that all of these colors will be moved to, or replaced with one in, a color in one of the other three files & what those files are.

The temporary sentence at the end sounds good.

BenHenning replied:
Please also add a TODO pointing to an issue to remove this file once the migration is complete.

ayush0402 replied:
Done.

BenHenning reviewed (commented):
Thanks @ayush0402! I didn't look too closely at the theme files since I'm not entirely sure how we're going to be managing those long-term, but the rest of the PR generally looks good (just had a few comments--PTAL).

@rt4914 how should these new colors relate to themes? Should we maybe be restricting themes to only reference color_palette colors similar to color_components?

ayush0402 committed:
used color_defs for color_palette instead of colors_migrating

ayush0402 committed:
added TODO in colors_migrating.xml

ayush0402 committed:
removed direct references to color_defs from component_colors.xml for static colors

ayush0402 commented:
@BenHenning PTAL.

rt4914 commented:
> Thanks @ayush0402! I didn't look too closely at the theme files since I'm not entirely sure how we're going to be managing those long-term, but the rest of the PR generally looks good (just had a few comments--PTAL).
> 
> @rt4914 how should these new colors relate to themes? Should we maybe be restricting themes to only reference color_palette colors similar to color_components?

@BenHenning yes agreed. Themes should connect to `color_palette.xml` only.

BenHenning reviewed (approved):
Thanks @ayush0402! I think the latest color files look really good, and I'm eager to see the rest of the colors get migrated over throughout the rest of the project. Thanks for your patience as we figured this out--I think it's moving toward much cleaner organization.

BenHenning commented:
Though to check one thing: @ayush0402 and @rt4914, is this actually finishing #4030 or is there more color migration work to do prior to dark mode being considered done?

ayush0402 commented:
@BenHenning No, we still have many views/layouts to check and cover before dark mode can be considered complete. I earlier had thought of filing this issue for setting the base for the implementation and then create more issues for other layouts and maybe making them available to community for some help. But i guess I will create more PRs under the same issue and have this PR title renamed to `Fix part of`. I have updated the issue description to contain proper explanation for the approach with links to important comments.
@rt4914 What are your views about getting the help of other contributors by filing issues after I make couple more PRs under this issue only for them to have proper references?

rt4914 commented:
> @BenHenning No, we still have many views/layouts to check and cover before dark mode can be considered complete. I earlier had thought of filing this issue for setting the base for the implementation and then create more issues for other layouts and maybe making them available to community for some help. But i guess I will create more PRs under the same issue and have this PR title renamed to `Fix part of`. I have updated the issue description to contain proper explanation for the approach with links to important comments. @rt4914 What are your views about getting the help of other contributors by filing issues after I make couple more PRs under this issue only for them to have proper references?

@ayush0402 Yes we can do that for sure. But I think we should create one more PR first and then do that. So that the new PR can act as sample PR and also we can be sure that everything is fully correct.
